File: php/148.php
<?php

/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val) { $this->val = $val; }
 * }
 */
class Solution {

    /**
     * @param ListNode $head
     * @return ListNode
     */
    function sortList($head) {
        if (!$head || !$head->next) {
            return $head;
        }
        $fast = $head->next;
        $slow = $head;
        while ($fast && $fast->next) {
            $fast = $fast->next->next;
            $slow = $slow->next;
        }
        $right = $slow->next;
        $slow->next = null;
        $left = $head;
        $left = $this->sortList($left);
        $right = $this->sortList($right);

        $dummyHead = new ListNode(null);
        $pre = $dummyHead;
        while ($left && $right) {
            if ($left->val < $right->val) {
                $pre->next = $left;
                $left = $left->next;
            } else {
                $pre->next = $right;
                $right = $right->next;
            }
            $pre = $pre->next;
        }
        $pre->next = $left ? $left : $right;
        return $dummyHead->next;
    }
}

File: go/linkedList/148.go
package main
//
//func sortList(head *ListNode) *ListNode {
//	if head == nil || head.Next == nil {
//		return head
//	}
//
//	fast, slow := head, head
//	fast = fast.Next
//	for fast != nil && fast.Next != nil {
//		fast = fast.Next.Next
//		slow = slow.Next
//	}
//
//	next := slow.Next
//	slow.Next = nil
//	left := sortList(head)
//	right := sortList(next)
//	return merge2LinkedList(left, right)
//}
//
//func merge2LinkedList(l1, l2 *ListNode) *ListNode {
//	if l1 == nil {
//		return l2
//	}
//	if l2 == nil {
//		return l1
//	}
//	if l1.Val > l2.Val {
//		l2.Next = merge2LinkedList(l1, l2.Next)
//		return l2
//	} else {
//		l1.Next = merge2LinkedList(l1.Next, l2)
//		return l1
//	}
//}
//
//type ListNode struct {
//	Val  int
//	Next *ListNode
//}
